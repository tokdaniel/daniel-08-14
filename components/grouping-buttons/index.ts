export { default as GroupingButtons } from './grouping-buttons'
