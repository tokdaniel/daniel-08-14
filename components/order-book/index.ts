export { default as OrderBook } from './order-book'
export { default as HalfBook } from './half-book'
