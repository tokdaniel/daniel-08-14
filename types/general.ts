export type Delta = [number, number][]
export type IncomingMessage = {
  event?: string
  feed: string
  product_ids: string[]
  asks?: Delta
  bids?: Delta
}

export type Book = {
  asks: Delta
  bids: Delta
}

export type Page = {
  [x: number]: number
}

export type NBook = {
  asks: Page
  bids: Page
}

export type Side = 'bids' | 'asks'
