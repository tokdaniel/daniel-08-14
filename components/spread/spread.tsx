import React from 'react'
import styled from 'styled-components'

import { GroupingButtons } from 'components/grouping-buttons'
import { Text } from 'components/typography'
import { NBook } from 'types/general'
import useSpread from 'util/hooks/use-spread'

const Container = styled.div`
  padding: 16px 8px;
  width: 100%;
`

const Header = styled.div`
  display: flex;
  margin-bottom: 12px;
`

const SpreadRow = styled.div`
  text-transform: uppercase;
  display: flex;
  align-items: baseline;
  justify-content: space-between;
  width: 100%;
`
type Props = {
  book: NBook
  grouping: number
  incrementGrouping: () => void
  decrementGrouping: () => void
}

const Spread: React.FC<Props> = ({
  book,
  grouping,
  incrementGrouping,
  decrementGrouping,
}) => {
  const { percentage, value, maxBid, minAsk } = useSpread(book)

  return (
    <Container>
      <Header>
        <Text variant="h5">SPREAD</Text>
        <GroupingButtons
          selected={grouping}
          increment={incrementGrouping}
          decrement={decrementGrouping}
        />
      </Header>
      <SpreadRow>
        <Text variant="body-2">percentage: </Text>
        <Text variant="body-1">{percentage}%</Text>
      </SpreadRow>
      <SpreadRow>
        <Text variant="body-2">value: </Text>
        <Text variant="body-1">{value} USD</Text>
      </SpreadRow>
      <SpreadRow>
        <Text variant="body-2">Highest Bid: </Text>
        <Text variant="body-1" textColor={['green', '500']}>
          {maxBid} USD
        </Text>
      </SpreadRow>
      <SpreadRow>
        <Text variant="body-2">Lowest Offer: </Text>
        <Text variant="body-1" textColor={['red', '500']}>
          {minAsk} USD
        </Text>
      </SpreadRow>
    </Container>
  )
}

export default Spread
