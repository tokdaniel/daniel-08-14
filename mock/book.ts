import { Book } from 'types/general'

export const simpleBook: Book = {
  bids: [
    [47599.5, 171],
    [47597, 1070],
    [47592, 4479],
  ],
  asks: [
    [47602.5, 2500],
    [47603, 1808],
    [47610, 2500],
  ],
}

export const book: Book = {
  bids: [
    [47599.5, 171],
    [47597, 1070],
    [47592, 4479],
    [47590.5, 170496],
    [47590, 5819],
    [47589.5, 2031],
    [47588, 2500],
    [47586, 26700],
    [47583, 11553],
    [47582.5, 10000],
    [47581.5, 4388],
    [47581, 3862],
    [47580, 57718],
    [47579.5, 3351],
    [47572.5, 6138],
    [47569.5, 4324],
    [47569, 62438],
    [47568.5, 203737],
    [47567, 10000],
    [47566, 2093],
    [47563, 10000],
    [47562.5, 20000],
    [47561.5, 10396],
    [47561, 11],
    [47559, 20000],
  ],
  asks: [
    [47602.5, 2500],
    [47603, 1808],
    [47610, 2500],
    [47610.5, 4043],
    [47613, 2500],
    [47614, 1986],
    [47614.5, 2500],
    [47615.5, 1070],
    [47617, 4637],
    [47617.5, 5315],
    [47622, 4565],
    [47622.5, 10000],
    [47623, 2500],
    [47623.5, 16138],
    [47626, 63948],
    [47626.5, 26700],
    [47627.5, 4737],
    [47628.5, 5000],
    [47629.5, 1976],
    [47637.5, 199658],
    [47638, 3026],
    [47638.5, 2982],
    [47641.5, 38081],
    [47642.5, 1937],
    [47644, 11365],
  ],
}
