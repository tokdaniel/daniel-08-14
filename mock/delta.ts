import { Book } from 'types/general'

export const delta: Book = {
  bids: [
    [47599.5, 0],
    [47597, 1259],
    [47592, 3345],
  ],
  asks: [
    [47602.5, 0],
    [47603, 2239],
    [47610, 1925],
  ],
}
