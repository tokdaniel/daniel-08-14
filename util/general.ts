import { Book, NBook, Page, Side } from 'types/general'

export const transformSide = (state: Book, side: 'bids' | 'asks'): Page => {
  return Object.values(state[side] ?? {}).reduce((acc, [price, size]) => {
    return {
      ...acc,
      [price]: size,
    }
  }, {})
}

export const transformBook = (book: Book): NBook => {
  const asks = transformSide(book, 'asks')
  const bids = transformSide(book, 'bids')

  return {
    asks,
    bids,
  }
}

export const sortBySide =
  (side: Side) =>
  ([priceA]: [string, number], [priceB]: [string, number]): number => {
    const a = parseFloat(priceA)
    const b = parseFloat(priceB)

    return side === 'asks' ? a - b : b - a
  }

export const computeDelta = (state: Page, delta: Page): Page =>
  Object.entries(Object.assign({}, state, delta))
    .filter(([_, size]) => size !== 0)
    .reduce(
      (acc, [price, size]) => ({
        ...acc,
        [price]: size,
      }),
      {},
    )

type Spread = {
  percentage: string
  value: number
}

export const calculateSpread = (minAsk: number, maxBid: number): Spread => {
  const value = minAsk - maxBid
  const percentage = (1 - minAsk / maxBid) * 100

  return { value, percentage: Math.abs(percentage).toFixed(3) }
}

export const getMinAsk = (asks: Page): number =>
  Math.min(...(Object.keys(asks) as unknown as number[]))

export const getMaxBid = (bids: Page): number =>
  Math.max(...(Object.keys(bids) as unknown as number[]))
