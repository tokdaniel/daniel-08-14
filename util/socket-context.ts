import { createContext, useContext } from 'react'
import { WebSocketSubject } from 'rxjs/webSocket'

import { IncomingMessage } from 'types/general'

const socketContext = createContext<{
  subject: WebSocketSubject<IncomingMessage> | null
}>({
  subject: null,
})

export const Provider = socketContext.Provider

export const useConnection = (): WebSocketSubject<IncomingMessage> => {
  return useContext(socketContext).subject
}
