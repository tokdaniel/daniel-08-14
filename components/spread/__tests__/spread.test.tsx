import React from 'react'
import renderer from 'react-test-renderer'
import { ThemeProvider } from 'styled-components'

import { Spread } from 'components/spread'
import theme from 'config/theme.config'
import { book } from 'mock/book'
import { transformBook } from 'util/general'

describe('Spread', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <ThemeProvider theme={theme}>
          <Spread
            book={transformBook(book)}
            grouping={25}
            incrementGrouping={() => null}
            decrementGrouping={() => null}
          />
          ,
        </ThemeProvider>,
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
