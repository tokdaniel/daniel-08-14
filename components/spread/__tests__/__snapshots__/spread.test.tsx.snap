// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spread renders correctly 1`] = `
Array [
  <div
    className="sc-hKFxyN WQCFw"
  >
    <div
      className="sc-eCApnc pJRi"
    >
      <div
        className="sc-bdnxRM ddgugd"
      >
        SPREAD
      </div>
      <div
        className="sc-gtsrHT dVvDsX"
      >
        <div
          className="sc-bdnxRM jDkucz"
        >
          group: 
          25
        </div>
        <button
          className="sc-dlnjwi gTMfzH"
          onClick={[Function]}
        >
          +
        </button>
        <button
          className="sc-dlnjwi gTMfzH"
          onClick={[Function]}
        >
          âˆ’
        </button>
      </div>
    </div>
    <div
      className="sc-jSFjdj bclzYn"
    >
      <div
        className="sc-bdnxRM kyahsn"
      >
        percentage: 
      </div>
      <div
        className="sc-bdnxRM cLcjEa"
      >
        0
        %
      </div>
    </div>
    <div
      className="sc-jSFjdj bclzYn"
    >
      <div
        className="sc-bdnxRM kyahsn"
      >
        value: 
      </div>
      <div
        className="sc-bdnxRM cLcjEa"
      >
        0
         USD
      </div>
    </div>
    <div
      className="sc-jSFjdj bclzYn"
    >
      <div
        className="sc-bdnxRM kyahsn"
      >
        Highest Bid: 
      </div>
      <div
        className="sc-bdnxRM jjbfsL"
      >
        0
         USD
      </div>
    </div>
    <div
      className="sc-jSFjdj bclzYn"
    >
      <div
        className="sc-bdnxRM kyahsn"
      >
        Lowest Offer: 
      </div>
      <div
        className="sc-bdnxRM cBeHvK"
      >
        0
         USD
      </div>
    </div>
  </div>,
  ",",
]
`;
